server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java

```diff
@@ -795,6 +1045,7 @@ public enum VectorSimilarity {
             float score(float similarity, ElementType elementType, int dim) {
                 return switch (elementType) {
                     case BYTE, FLOAT -> 1f / (1f + similarity * similarity);
+                    case BIT -> dim - (similarity * dim);
```

Comment:
I am confused about this. What does`similarity` represent here?  Is it the number of similar bits?  Shouldn't than the score be: ` similarity/dim`?

Reply:
The hamming score transformation is:
```
((a.length * Byte.SIZE) - VectorUtil.xorBitCount(a, b)) / (float) (a.length * Byte.SIZE);
```

`a.length * Byte.SIZE` is just the `dim` here as in the mapping we account for it representing bits. In Lucene, we account for the `byte` values (thus we need to correct it). 

As for if this is the correct way to transform the score. Let's take it case by case. Really we only have two cases, `dotProduct` (which covers cosine, max-inner-product), and `euclidean`. 

For `euclidean`, the `l2norm` distance is just `hamming`. But with `l2norm` directly, values that are higher are further away. So we must do a scoring inversion. So, doing `bitDimensionCount - xor` gives us the number of bits that are the same (inverting the l2norm so that bigger values give us a smaller score, e.g. vectors further away score lower). I then normalize by `dim` to scale us similarly with all our other vector scores (the max possible _score for bit vectors is `dim` as we just count the number of bits that are the same).


For dotProduct, surprisingly, the same transformation works. But it has the assumption that the provided vectors are bit-quantized (meaning transformed to `0` and `1` around some median, and thus having the exact same bits in that case is beneficial).


Maybe if we have `l2norm` be the inverted `xor` score, we should consider making `dotProduct` an actual `dotProduct` (a `bit AND` score). I did this originally. 

@tveasey @jimczi what do you think? How should we handle dotProduct (and its family of max inner product and cosine)?

Reply:
As discussed offline there are subtleties around this that I'm just noting here for completeness.

The crux of the matter is how the vectors are quantised. Assuming you want minimum mean square error estimates and you quantise as x_i <= median_i |-> 0 and x_i > median |-> 1 then one can work through minimum mean squared expected estimators for each type of distance / similarity function. Generally, I think people are assuming each component is IID (which usually isn't too bad for learned representations). This allows you to compute distances/similarities directly on bit vectors up to some constant scaling.

Taking L2 as a prototype we have

dist(0, 0) = (E[x | x <= median] - E[x | x <= median])^2 = 0 (similar for dist(1, 1))
dist(0, 1) = (E[x | x > median] - E[x | x <= median])^2 = k^2

so our best estimate for L2  would be k^2 * count(x ^ y). (One can run a similar sort of analysis for dot.)

Reply:
@benwtrent @tveasey Thanks for detailed explanations.

@benwtrent If similarity in this funciton is ` hamming`, why `score` is not calculated as you defined in documentation: `(numBits - hamming(a, b)) / numBits` , so will be 

`(dim - similarity) / dim` ?

Reply:
@mayya-sharipova you are correct, let me fix it!